xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -170,6 +187,59 @@ void watchConfigData(String hostName, int port, ConfigWatcher watcher) {
     adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
   }
 
+  @Override
+  void watchEndpointData(String clusterName, EndpointWatcher watcher) {
+    boolean needRequest = false;
+    if (!endpointWatchers.containsKey(clusterName)) {
+      logger.log(Level.FINE, "Start watching endpoints in cluster {0}", clusterName);
+      needRequest = true;
+      endpointWatchers.put(clusterName, new HashSet<EndpointWatcher>());
+    }
+    Set<EndpointWatcher> watchers = endpointWatchers.get(clusterName);
+    if (watchers.contains(watcher)) {
+      logger.log(Level.WARNING, "Watcher {0} already registered", watcher);
+    }
```

Comment:
The above can be modified as follows (which I prefer):
```
    Set<EndpointWatcher> watchers = null;
    if (!endpointWatchers.containsKey(clusterName)) {
      logger.log(Level.FINE, "Start watching endpoints in cluster {0}", clusterName);
      needRequest = true;
      watchers = new HashSet<EndpointWatcher>();
      endpointWatchers.put(clusterName, watchers);
    } else {
      watchers = endpointWatchers.get(clusterName);
      if (watchers.contains(watcher)) {
         logger.log(Level.WARNING, "Watcher {0} already registered", watcher);
      }
    }
```

Reply:
Hmm... Personal styling preference. I usually would like to minimize variable reassignments so that their values are easier to follow. But whatever, we may want to spend less time on arguing this as long as there is no sever logical bug.